syntax = "proto3";

package dutTestAgent;

option go_package = "dut_test_agent/proto/dut_agent";

message TestRequest {
  string device_id = 1;
  map<string, string> test_params = 2; // Specific parameters for the test, e.g., "test_duration": "5"
}

message TestResponse {
  bool passed = 1;
  string message = 2;
  map<string, string> details = 3; // Detailed key-value pairs of results
}

message TestInfo {
  string name = 1;         // The unique name of the test (e.g., "RunAccelTest").
  string description = 2;  // A brief description of what the test does.
}

message GetAvailableTestsRequest {}

message GetAvailableTestsResponse {
  repeated TestInfo available_tests = 1; // List of tests available for execution.
}

message GetAgentDetailsRequest {}

message GetAgentDetailsResponse {
  string agent_version = 1; // The version of the agent software.
  string hostname = 2;      // The hostname of the machine where the agent is running.
  string device_id = 3;     // The ID of the Device Under Test (DUT) the agent is currently connected to.
}

service DutAgentService {
  // GetAgentDetails retrieves basic identification and status information about the agent.
  // This can be used as a simple health check and to verify agent connectivity.
  rpc GetAgentDetails (GetAgentDetailsRequest) returns (GetAgentDetailsResponse);

  // RunCommand executes a shell command on the DUT.
  rpc RunCommand (RunCommandRequest) returns (RunCommandResponse);

  // GetDeviceInfo retrieves basic identification and status information about the DUT.
  rpc GetDeviceInfo (GetDeviceInfoRequest) returns (GetDeviceInfoResponse);

  // StreamExecutionLogs enables server-side streaming of execution logs from the agent to the client.
  rpc StreamExecutionLogs (StreamExecutionLogsRequest) returns (stream LogEntry);
  
  // RunTest executes a specific hardware test on the DUT.
  // The type of test and its parameters are specified within the TestRequest.
  rpc RunTest(TestRequest) returns (TestResponse);

  // GetAvailableTests retrieves a list of hardware tests that can be executed on the DUT.
  rpc GetAvailableTests (GetAvailableTestsRequest) returns (GetAvailableTestsResponse);
}

message RunCommandRequest {
  string command = 1;          // The command to execute (e.g., "ls", "adb shell getprop").
  repeated string args = 2;    // Arguments to pass to the command.
  int32 timeout_seconds = 3;   // Timeout for the command execution in seconds.
  bool sudo = 4;               // If true, attempt to run the command with superuser privileges.
  string working_directory = 5; // (Optional) Working directory for the command.
}

message RunCommandResponse {
  int32 exit_code = 1;       // The exit code of the executed command.
  string stdout = 2;         // Standard output from the command.
  string stderr = 3;         // Standard error from the command.
  string error_message = 4;  // Any error message generated by the agent itself (e.g., timeout, command not found).
}

message GetDeviceInfoRequest {}

message GetDeviceInfoResponse {
  string hostname = 1;      // The hostname of the DUT.
  string os_name = 2;       // e.g., "Linux", "Android".
  string os_version = 3;    // e.g., "13", "5.10.0-10-android".
  string architecture = 4;  // e.g., "arm64", "x86_64".
  // A map for any other arbitrary key-value device properties.
  map<string, string> custom_fields = 5; 
}

enum LogLevel {
  LOG_LEVEL_UNSPECIFIED = 0;
  LOG_LEVEL_DEBUG = 1;
  LOG_LEVEL_INFO = 2;
  LOG_LEVEL_WARN = 3;
  LOG_LEVEL_ERROR = 4;
  LOG_LEVEL_FATAL = 5;
}

message StreamExecutionLogsRequest {
  LogLevel min_level = 1; // Only stream logs with this level or higher severity.
  string component_filter = 2; // (Optional) Filter logs by a specific component name.
  bool follow = 3; // If true, stream new logs as they are generated; otherwise, send historical logs and close.
}

message LogEntry {
  string timestamp = 1; // ISO 8601 formatted timestamp, e.g., "2006-01-02T15:04:05Z07:00"
  LogLevel level = 2;
  string component = 3; // (Optional) The component or module that generated the log.
  string message = 4;
}